### PR TITLE
Dem feature parallelize rigid elements

### PR DESCRIPTION
**Pull request by joaquinirazabal**
The parallelism in the function GetRigidBodyElementsForce was applied over the elements loop. As the rigid elements used to be  a few with a lot of nodes, now the parallel loop is applied over the nodes, as it was discussed with @maceligueta.

### CODE REVIEW

**Review by GuillermoCasas — CHANGES_REQUESTED** [review]
should center_to_node_vector not be made private to the thread (idem with additional_torque)?

**Comment by joaquinirazabal** [comment]
Of course, @GuillermoCasas, thank you for the feedback. I've just pushed the changes you suggested.

**Comment by maceligueta** [comment]
Hi, don't you think this style is too long?
```
vector<unsigned int> node_partition;
array_1d<double, 3> additional_torque;		 
int NumberOfThreads = OpenMPUtils::GetNumThreads();        
OpenMPUtils::CreatePartition(NumberOfThreads, mListOfNodes.size(), node_partition);  		  
NodesArrayType::iterator i_begin = mListOfNodes.begin();		
#pragma omp parallel for
NodesArrayType::iterator i_end = mListOfNodes.end();		
for (int k = 0; k < NumberOfThreads; k++) {
       NodesArrayType::iterator i_begin = mListOfNodes.begin() + node_partition[k];
        for (ModelPart::NodeIterator i = i_begin; i != i_end; ++i) {		 
            NodesArrayType::iterator i_end = mListOfNodes.begin() + node_partition[k + 1];
```
I am trying to move to something like:

```
#pragma omp parallel for
for (int k = 0; k < mListOfNodes.size(); k++) {
   ModelPart::NodeIterator i = mListOfNodes.begin() + k;
    ...
```
open for discussion, but the second one allows different types of scheduling, while the first one is always static (apart from a really big chunk of code).

**Comment by joaquinirazabal** [comment]
I agree with you @maceligueta, the way you suggested is more elegant. I changed it in RigidBodyElement and also in the strategy were the ugly chunk of code appeared twice.

**Comment by maceligueta** [comment]
@joaquinirazabal , many thanks! 
Concerning the 'schedule dynamic(100)', I use it when I expect the loop to be unbalanced. However, for loops where the work is exactly the same for all the iterations, I don't put it. For instance: spheres can have different number of neighbours, so the amount of work is different for each sphere, then I use dynamic scheduling. For the integration of trajectories or for setting values to 0.0 I would not use it, because it has an overhead that would bring no benefit...

**Comment by maceligueta** [comment]
@joaquinirazabal , I made some slight changes that -I hope- did not break anything. Can you try your specific test?

**Comment by joaquinirazabal** [comment]
It works!!

**Review by maceligueta — CHANGES_REQUESTED** [review]
I just saw that the RigidBodyElement has one little problem: when it's a wall, the parallel loop should be over its nodes (few walls, many nodes), however, when it's a cluster the parallel loop should be over the clusters (many clusters, few nodes per cluster). It's just for performance...
The solution we can manufacture is to create the object ParallelizedRigidBody3D, where all loops are parallel. Then, walls would derive from the parallel version, while the clusters would derive from the serial version. 
@pooyan-dadvand (or anyone) could bring ideas to this maybe...

File: applications/DEM_application/custom_elements/rigid_body_element.cpp [hunk 1]
```diff
@@ -200,7 +200,7 @@ namespace Kratos {
         center_forces[0] = center_forces[1] = center_forces[2] = center_torque[0] = center_torque[1] = center_torque[2] = 0.0;
 
         #pragma omp parallel for schedule(dynamic, 100)
-        for (unsigned int k = 0; k < mListOfNodes.size(); k++) {
+        for (int k = 0; k < (int)mListOfNodes.size(); k++) {
```
Comment (pooyan-dadvand) [inline applications/DEM_application/custom_elements/rigid_body_element.cpp]: I would suggest the use of static_cast for casting:

```c++
for (int k = 0; k < static_cast<int>(mListOfNodes.size()); k++) {
```
The reasoning is the fact that now the casting is more visible and is obvious that a compile time casting is done.

**Comment by pooyan-dadvand** [comment]
I understand that you have a parallel loop over these loops you are mentioning. Then I would just make them with dynamic scheduling and leave the cluster loops all in serial avoiding nested parallelization. Maybe @RiccardoRossi has a better solution.

**Comment by maceligueta** [comment]
What about this?
```
#pragma omp parallel for if(number_of_solids > 100)
for (int i = 0; i < number_of_solids; k++) {
    #pragma omp parallel for if(number_of_nodes_per_solid > 300 && number_of_solids <= 100)
    for (int k = 0; k < number_of_nodes_per_solid; k++) {
```

**Comment by maceligueta** [comment]
Apparently, my last idea is not working at all. It seems like for a small number_of_solids OpenMP decides to work in one thread, then the next #pragma (over nodes of the solid) tries to make the parallelism inside that single thread, so it gets really slow. I tried it with a dummy code.
This makes me think that the apporach proposed by @pooyan-dadvand with the 'dynamic' clause would lead to the same bad result.

**Comment by pooyan-dadvand** [comment]
In my approach the inner loop is only in serial to avoid the nested parallel loop problem you got in your example. The dynamic would give benefit if you have a large number of clusters with heterogeneous loop sizes.

**Comment by maceligueta** [comment]
I just found out that the loop over the walls and the loop over the clusters are totally separated, so the outer loop is already specialized. The inner loop, right now, is also independent, so we have to find some way to unify them (or not, seeing the problems that this might bring). This will be in a future PR.

**Review by maceligueta — APPROVED** [review]

**Comment by maceligueta** [comment]
@GuillermoCasas:  approval?

**Review by GuillermoCasas — APPROVED** [review]

**Pull request merged**